### PR TITLE
SHOP-2375, SHOP-2376: Reference Transactions

### PR DESCRIPTION
**Pull request by laithalissa**
Add functionality to allow GMGPay to make billing agreements and
perform reference transactions.
- Step up API version from 53.0 to 58.0. 58.0 is the minimum version
  which supports
    L_BILLINGTYPEn: MerchantInitiatedBillingSingleAgreement
  (Paypal recommends MIBSA in the statement of work for GMG)
- Add UpdateAuthorization method as a wrapper to update billing
  agreements, or to check a billing agreement's validity
- Add DoReferenceTransaction to make reference transactions using
  a billing agreement (name changes to referenceid in this context)
- Refactor some shared functionality into a private method, only
  methods which share the code verbatim are affected.

### CODE REVIEW

File: paypal/driver.py [hunk 1]
```diff
@@ -117,6 +121,30 @@ def _get_value_from_qs(self, qs, value):
...
+        :type parameters: dictionary
+
+        :returns: value of the ACK parameter in the response, can be
+            "Success", "SuccessWithWarning", "Failure", "FailureWithWarning"
+        :rtype: string
+        """
+        query_string = self.signature + urllib.urlencode(parameters or {})
+        response = urllib2.urlopen(self.NVP_API_ENDPOINT, query_string).read()
+        response_tokens = {}
+        for token in response.split('&'):
```
Comment (giftig) [inline paypal/driver.py:137]: I don't like this but it's the original library code so meh.

**Comment by giftig** [comment]
LGTM.

**Comment by DSmedley1989** [comment]
image

**Comment by giftig** [comment]
For Adam's sake, LGTM means let's ship this please.

**Pull request merged**